### PR TITLE
Release search context when scroll keep_alive is too large

### PR DESCRIPTION
**Pull request by dnhatn**
Previously, we close related search contexts if the keep_alive of a scroll is too large.  But we accidentally change this behavior in #62061.

CI: https://gradle-enterprise.elastic.co/s/f2aczys5rxvik/console-log/raw

### CODE REVIEW

**Comment by dnhatn** [comment]
@jimczi I've updated the PR.

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 1]
```diff
@@ -471,7 +471,14 @@ public void executeQueryPhase(InternalScrollSearchRequest request,
                                   SearchShardTask task,
                                   ActionListener<ScrollQuerySearchResult> listener) {
         final LegacyReaderContext readerContext = (LegacyReaderContext) findReaderContext(request.contextId(), request);
-        final Releasable markAsUsed = readerContext.markAsUsed(getScrollKeepAlive(request.scroll()));
+        final Releasable markAsUsed;
+        try {
+            markAsUsed = readerContext.markAsUsed(getScrollKeepAlive(request.scroll()));
+        } catch (Exception e) {
+            // We need to release the reader context of the scroll when we hit any exception (here the keep_alive can be too large)
+            processFailure(readerContext, e);
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/SearchService.java:479]: can you replace the call with `freeReaderContext(context.id())`?

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 2]
```diff
@@ -536,7 +543,14 @@ private Executor getExecutor(IndexShard indexShard) {
     public void executeFetchPhase(InternalScrollSearchRequest request, SearchShardTask task,
                                   ActionListener<ScrollQueryFetchSearchResult> listener) {
         final LegacyReaderContext readerContext = (LegacyReaderContext) findReaderContext(request.contextId(), request);
-        final Releasable markAsUsed = readerContext.markAsUsed(getScrollKeepAlive(request.scroll()));
+        final Releasable markAsUsed;
+        try {
+            markAsUsed = readerContext.markAsUsed(getScrollKeepAlive(request.scroll()));
+        } catch (Exception e) {
+            // We need to release the reader context of the scroll when we hit any exception (here the keep_alive can be too large)
+            processFailure(readerContext, e);
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/SearchService.java:551]: same here

**Review by jimczi — APPROVED** [review]
I left a small comments but the fix looks good to me.

File: server/src/main/java/org/elasticsearch/search/SearchService.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/search/SearchService.java]: ++. I pushed https://github.com/elastic/elasticsearch/pull/62179/commits/3517cb56f1130288b92b9587e65e267fa6873626

**Review by jimczi — APPROVED** [review]

**Comment by dnhatn** [comment]
Thanks Jim!

**Pull request merged**